tests/lib/rules/no-redeclare.js

```diff
@@ -739,3 +739,592 @@ ruleTester.run("no-redeclare", rule, {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a = function () {};
+	  var a = new Date();
+			`,
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a = 3;
+	  var a = 10;
+	  var a = 15;
+			`,
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var a;
+			`,
+			languageOptions: { parserOptions: { sourceType: "module" } },
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  export var a;
+	  var a;
+			`,
+			languageOptions: { parserOptions: { sourceType: "module" } },
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: "var Object = 0;",
+			options: [{ builtinGlobals: true }],
+			errors: [
+				{
+					data: {
+						id: "Object",
+					},
+					messageId: "redeclaredAsBuiltin",
+				},
+			],
+		},
+		{
+			code: "var top = 0;",
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				globals: { top: "readonly" },
+			},
+			errors: [
+				{
+					data: {
+						id: "top",
+					},
+					messageId: "redeclaredAsBuiltin",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: { parserOptions: { ecmaVersion: 6 } },
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+				{
+					data: {
+						id: "Object",
+					},
+					messageId: "redeclaredAsBuiltin",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: { ecmaVersion: 6, sourceType: "module" },
+			},
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: true }],
+			languageOptions: {
+				parserOptions: {
+					ecmaFeatures: { globalReturn: true },
+					ecmaVersion: 6,
+				},
+			},
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+		{
+			code: `
+	  var a;
+	  var { a = 0, b: Object = 0 } = {};
+			`,
+			options: [{ builtinGlobals: false }],
+			languageOptions: { parserOptions: { ecmaVersion: 6 } },
+			errors: [
+				{
+					data: {
+						id: "a",
+					},
+					messageId: "redeclared",
+				},
+			],
+		},
+
+		// Notifications of readonly are moved from no-undef: https://github.com/eslint/eslint/issues/4504
+		{
+			code: "/*global b:false*/ var b = 1;",
+			options: [{ builtinGlobals: true }],
```

Comment:
This case isn't reported in JavaScript ([Playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVkZWNsYXJlOiBbXCJlcnJvclwiLCB7IFwiYnVpbHRpbkdsb2JhbHNcIjogdHJ1ZSB9XSovXG5cbi8qZ2xvYmFsIGI6ZmFsc2UqLyB2YXIgYiA9IDE7Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)). I don't see why it should be reported in TypeScript.